docs/spec_draft.md

```diff
@@ -522,6 +523,83 @@ it only exists to establish data dependencies from `result` to `inputs`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups).
+
+Afterwards, within each `process_group`:
+  * `operands@receiver = [operand@sender for sender in process_group]` for all
+    `receiver` in `process_group`.
+  * ```
+    result@process[i0, i1, ..., iR-1 =
+        reduce(
+          inputs=[operands@process[j][i0, i1, ..., iR-1] for j in range(size(operands@process))],
+          init_values=[I, such that `computation` and `I` form a monoid],
+          dimensions=[0],
+          body=computation
+        )
+    ```
+
+### Inputs
+
+| Name                    | Type                                         |
+|-------------------------|----------------------------------------------|
+| `operand`               | tensor of any supported type                 |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64` |
+| `channel_id`            | constant of type `si64`                      |
+| `use_global_device_ids` | constant of type `boolean`                   |
+| `computation`           | `function`                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) size(`replica_groups`) $\gt$ 0.
+  * (C2) All values in `replica_groups` are unique.
+  * (C3) `size(replica_groups)` = `num_replicas`.
+  * (C4) $0 \le$ `replica_groups`[i] $\lt$ size(`replica_groups`) $\forall i$
```

Comment:
can we use replica_id here, since the doc already outlines it.

Reply:
WDYT?

Reply:
Maybe I do not understand the point.
Here we are verifying the `replica_groups` parameter attached to the instruction.

Reply:
is C4 is trying to say the following?
0 $\le$ `replica_id  $\lt$ `num_replicas`

Reply:
C3 and C4 together mean - `replica_groups` cover all `replica_ids` once and exactly once.

Reply:
"C3 and C4 together mean - replica_groups cover all replica_ids once and exactly once". The wording here follows similar wording in gather, scatter, etc, although indeed there are multiple ways to express the same constraint.

What we've been often doing in the past when encountering situations when there are multiple possible formulations for some formalism is making a note in #289 and proceeding as is. This way, we'll be able to optimize various pieces of formalism globally, once the spec is complete, rather than potentially oscillating in individual PRs.